src/profiling/profiler.hpp

```diff
@@ -1,255 +1,56 @@
-    region_type& regions() { return root_region_; }
-
-    /// return a pointer to the current region
-    region_type* current_region() { return current_region_; }
-
-    /// return if in the root region (i.e. the highest level)
-    bool is_in_root() const { return &root_region_ == current_region_; }
-
-    /// return if the profiler has been activated
-    bool is_activated() const { return activated_; }
-
-    /// start (activate) the profiler
-    void start();
-
-    /// stop (deactivate) the profiler
-    void stop();
-
-    /// restart the profiler
-    /// remove all trace information and restart timer for the root region
-    void restart();
-
-    /// the time stamp at which the profiler was started (avtivated)
-    timer_type::time_point start_time() const { return start_time_; }
-
-    /// the time stamp at which the profiler was stopped (deavtivated)
-    timer_type::time_point stop_time()  const { return stop_time_; }
-
-    /// the time in seconds between activation and deactivation of the profiler
-    double wall_time() const {
-        return timer_type::difference(start_time_, stop_time_);
-    }
-
-    /// stop the profiler then generate the performance tree ready for output
-    profiler_node performance_tree();
-
-private:
-    void activate()   { activated_ = true;  }
-    void deactivate() { activated_ = false; }
-
-    timer_type::time_point start_time_;
-    timer_type::time_point stop_time_;
-    bool activated_ = false;
-    region_type root_region_;
-    region_type* current_region_ = &root_region_;
-};
-
-#ifdef ARB_HAVE_PROFILING
-namespace data {
-    using profiler_wrapper = arb::threading::enumerable_thread_specific<profiler>;
-    extern profiler_wrapper profilers_;
-}
-#endif
-
-/// get a reference to the thread private profiler
-/// will lazily create and start the profiler it it has not already been done so
-profiler& get_profiler();
-
-/// start thread private profiler
 void profiler_start();
-
-/// stop thread private profiler
 void profiler_stop();
-
-/// enter a profiling region with name n
-void profiler_enter(const char* n);
-
-/// enter nested profiler regions in a single call
-template <class...Args>
-void profiler_enter(const char* n, Args... args) {
-#ifdef ARB_HAVE_PROFILING
-    get_profiler().enter(n);
-    profiler_enter(args...);
-#endif
-}
-
-/// move up one level in the profiler
+void profiler_restart();
+void profiler_enter(std::size_t region_id);
 void profiler_leave();
 
-/// move up multiple profiler levels in one call
-void profiler_leave(int nlevels);
+profile profiler_summary();
+std::size_t profiler_region_id(const char* name);
+void profiler_print(const profile& prof, float threshold=1);
 
-/// iterate and stop them
-void profilers_stop();
+// enter a profiling region
+#define PE(name) \
+    static std::size_t name ## _profile_region_tag__ = arb::util::profiler_region_id(#name); \
+    arb::util::profiler_enter(name ## _profile_region_tag__);
```

Comment:
An alternative to using the region name as the key would be to use the `__COUNTER__` preprocessor macro to generate unique names. It's not standard, but it's quite portable across compilers. This would also obviate the need for the `PR` macro, and we could use a separator like `/` in region names, if that appears more natural.

Reply:
I am not very comfortable with using `__COUNTER__`, because it is non-standard... but it is tempting.

Reply:
We do also use `#pragma once` :) Not sure though of their respective spans of support across compilers. With the limitation of sticking to one `PE` per line, using `__LINE__` is a possible fallback.

Reply:
Hmm. Everything (including MSVC) on Godbolt supports it, as does xlc v. 13 at least.

Reply:
I have implemented it. Seems to work pretty well... we can roll it back if it causes problems in the future.